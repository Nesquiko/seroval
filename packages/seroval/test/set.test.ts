/* eslint-disable no-await-in-loop */
import { describe, it, expect } from 'vitest';
import {
  AsyncServerValue,
  deserialize,
  Feature,
  fromJSON,
  serialize,
  serializeAsync,
  ServerValue,
  toJSON,
  toJSONAsync,
} from '../src';

describe('Set', () => {
  describe('serialize', () => {
    it('supports Set', () => {
      const example = new Set([1, 2, 3]);
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Set<number>>(result);
      expect(back).toBeInstanceOf(Set);
      expect(back.has(1)).toBe(example.has(1));
      expect(back.has(2)).toBe(example.has(2));
      expect(back.has(3)).toBe(example.has(3));
    });
    it('supports self-recursion', () => {
      const example: Set<ServerValue> = new Set();
      example.add(example);
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Set<ServerValue>>(result);
      expect(back.has(back)).toBe(true);
    });
  });
  describe('serializeAsync', () => {
    it('supports Set', async () => {
      const example = new Set([1, 2, 3]);
      const result = await serializeAsync(Promise.resolve(example));
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<Set<number>>>(result);
      expect(back).toBeInstanceOf(Set);
      expect(back.has(1)).toBe(example.has(1));
      expect(back.has(2)).toBe(example.has(2));
      expect(back.has(3)).toBe(example.has(3));
    });
    it('supports self-recursion', async () => {
      const example: Set<AsyncServerValue> = new Set();
      example.add(Promise.resolve(example));
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Set<Promise<any>>>(result);
      for (const key of back) {
        expect(await key).toBe(back);
      }
    });
  });
  describe('toJSON', () => {
    it('supports Set', () => {
      const example = new Set([1, 2, 3]);
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Set<number>>(result);
      expect(back).toBeInstanceOf(Set);
      expect(back.has(1)).toBe(example.has(1));
      expect(back.has(2)).toBe(example.has(2));
      expect(back.has(3)).toBe(example.has(3));
    });
    it('supports self-recursion', () => {
      const example: Set<ServerValue> = new Set();
      example.add(example);
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Set<ServerValue>>(result);
      expect(back.has(back)).toBe(true);
    });
  });
  describe('toJSONAsync', () => {
    it('supports Set', async () => {
      const example = new Set([1, 2, 3]);
      const result = await toJSONAsync(Promise.resolve(example));
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<Set<number>>>(result);
      expect(back).toBeInstanceOf(Set);
      expect(back.has(1)).toBe(example.has(1));
      expect(back.has(2)).toBe(example.has(2));
      expect(back.has(3)).toBe(example.has(3));
    });
    it('supports self-recursion', async () => {
      const example: Set<AsyncServerValue> = new Set();
      example.add(Promise.resolve(example));
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Set<Promise<any>>>(result);
      for (const key of back) {
        expect(await key).toBe(back);
      }
    });
  });
  describe('compat', () => {
    it('should throw an error for unsupported target', () => {
      expect(() => serialize(new Set([1, 2, 3]), {
        disabledFeatures: Feature.Set,
      })).toThrowErrorMatchingSnapshot();
    });
  });
  describe('compat#toJSON', () => {
    it('should throw an error for unsupported target', () => {
      expect(() => toJSON(new Set([1, 2, 3]), {
        disabledFeatures: Feature.Set,
      })).toThrowErrorMatchingSnapshot();
    });
  });
});
