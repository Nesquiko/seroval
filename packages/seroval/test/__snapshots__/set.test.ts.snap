// Vitest Snapshot v1

exports[`Set > compat > should throw an error for unsupported target 1`] = `"Unsupported type \\"Set\\""`;

exports[`Set > compat#toJSON > should throw an error for unsupported target 1`] = `"Unsupported type \\"Set\\""`;

exports[`Set > serialize > supports Set 1`] = `"new Set([1,2,3])"`;

exports[`Set > serialize > supports self-recursion 1`] = `"(h=>(h=new Set,h.add(h),h))()"`;

exports[`Set > serializeAsync > supports Set 1`] = `"Promise.resolve(new Set([1,2,3]))"`;

exports[`Set > serializeAsync > supports self-recursion 1`] = `"(h=>(h=new Set([Promise.resolve().then(()=>h)])))()"`;

exports[`Set > toJSON > supports Set 1`] = `"{\\"t\\":{\\"t\\":13,\\"i\\":0,\\"l\\":3,\\"a\\":[{\\"t\\":0,\\"s\\":1},{\\"t\\":0,\\"s\\":2},{\\"t\\":0,\\"s\\":3}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`Set > toJSON > supports self-recursion 1`] = `"{\\"t\\":{\\"t\\":13,\\"i\\":0,\\"l\\":1,\\"a\\":[{\\"t\\":10,\\"i\\":0}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[0]}"`;

exports[`Set > toJSONAsync > supports Set 1`] = `"{\\"t\\":{\\"t\\":18,\\"i\\":0,\\"f\\":{\\"t\\":13,\\"i\\":1,\\"l\\":3,\\"a\\":[{\\"t\\":0,\\"s\\":1},{\\"t\\":0,\\"s\\":2},{\\"t\\":0,\\"s\\":3}]}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`Set > toJSONAsync > supports self-recursion 1`] = `"{\\"t\\":{\\"t\\":13,\\"i\\":0,\\"l\\":1,\\"a\\":[{\\"t\\":18,\\"i\\":1,\\"f\\":{\\"t\\":10,\\"i\\":0}}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[0]}"`;
